File: blog-backend/blog-core/src/main/java/com/zach/blog/service/RoleService.java
package com.zach.blog.service;

import com.zach.blog.dto.request.CreateRoleRequest;
import com.zach.blog.dto.request.UpdateRoleRequest;
import com.zach.blog.model.Role;
import org.springframework.data.domain.Page;

import java.util.List;

public interface RoleService {

    Role findOrCreateRole(String roleName);

    Page<Role> getRoles(Integer pageNum, Integer pageSize, String roleName, Boolean enable);

    void changeRoleStatus(Long id, boolean enable);

    Role createRole(CreateRoleRequest createRoleRequest);

    void updateRole(Long id, UpdateRoleRequest request);

    void deleteRole(Long id);

    List<Role> getAllActiveRoles();

    Role getRoleDetails(Long roleId);
}

File: blog-frontend/blog-admin/src/services/resources/role-service.ts
import { resourceAxios } from "../../utils/axios-utils"

export const roleService = {
    getRoles: async ({ pageSize = 5, pageNum = 0, name, enable }: GetRoles) => {
        const result = await resourceAxios.get<void, PageRespsone<Role>>('/roles', {
            params: {
                pageSize,
                pageNum,
                ...(name !== '' ? { roleName: name } : {}),
                ...(enable !== null ? { enable } : {})
            }
        })
        return result
    },

    getAllActiveRoles: async () => {
        return await resourceAxios.get<void, RoleNameResponse[]>('/roles/all')
    },

    getRoleDetails: async (roleId: number) => {
        return await resourceAxios.get<void, EditRole>(`/roles/${roleId}`)
    },

    changeRoleStatus: async ({ id, enable }: ChangeStatusRequest) => {
        return await resourceAxios.put<void, void>(`/roles/${id}/status`, {
            enable
        })
    },

    createRole: async (request: CreateRoleRequest) => {
        return await resourceAxios.post<void, Role>('/roles', request)
    },

    updateRole: async (request: UpdateRoleRequest) => {
        const { id: roleId } = request
        const result = await resourceAxios.put<void, void>(`/roles/${roleId}`, {
            ...request
        })
        return result
    },

    deleteRole: async (roleId: number) => {
        await resourceAxios.delete<void, void>(`/roles/${roleId}`)
    }
}